File: src/Fs.Units.CSharp/Simple/Temperature.cs
// ReSharper disable once CheckNamespace
namespace Fs.Units.CSharp.Temperature;

public static class KelvinExtensions
{
    public static double KelvinToCelsius(this double x) => Units.Temperature.Kelvin.toCelsius(x);
    public static double KelvinToFahrenheit(this double x) => Units.Temperature.Kelvin.toFahrenheit(x);
    public static double KelvinToRankine(this double x) => Units.Temperature.Kelvin.toRankine(x);
}

public static class CelsiusExtensions
{
    public static double CelsiusToKelvin(this double x) => Units.Temperature.Celsius.toKelvin(x);
    public static double CelsiusToFahrenheit(this double x) => Units.Temperature.Celsius.toFahrenheit(x);
    public static double CelsiusToRankine(this double x) => Units.Temperature.Celsius.toRankine(x);
}

public static class FahrenheitExtensions
{
    public static double FahrenheitToKelvin(this double x) => Units.Temperature.Fahrenheit.toKelvin(x);
    public static double FahrenheitToCelsius(this double x) => Units.Temperature.Fahrenheit.toCelsius(x);
    public static double FahrenheitToRankine(this double x) => Units.Temperature.Fahrenheit.toRankine(x);
}

public static class RankineExtensions
{
    public static double RankineToKelvin(this double x) => Units.Temperature.Rankine.toKelvin(x);
    public static double RankineToCelsius(this double x) => Units.Temperature.Rankine.toCelsius(x);
    public static double RankineToFahrenheit(this double x) => Units.Temperature.Rankine.toFahrenheit(x);
}

File: src/Fs.Units/Simple/Temperature.fs
namespace Fs.Units

[<AutoOpen>]
module Temperature =

  /// Kelvin
  [<Measure>]
  type K

  /// Celsius
  [<Measure>]
  type C

  /// Fahrenheit
  [<Measure>]
  type F

  /// Rankine
  [<Measure>]
  type R

  [<RequireQualifiedAccess>]
  module Kelvin =

    let create (x: float) = x * 1.0<K>
    let toCelsius (x: float<K>) = (x - 273.15<K>) / 1.0<K> * 1.0<C>

    let toFahrenheit (x: float<K>) =
      ((x - 273.15<K>) * (9.0 / 5.0) + 32.0<K>) / 1.0<K> * 1.0<F>

    let toRankine (x: float<K>) = (x * 1.8) / 1.0<K> * 1.0<R>

  [<RequireQualifiedAccess>]
  module Celsius =

    let create (x: float) = x * 1.0<C>
    let toKelvin (x: float<C>) = (x + 273.15<C>) / 1.0<C> * 1.0<K>

    let toFahrenheit (x: float<C>) =
      ((x * (9.0 / 5.0)) + 32.0<C>) / 1.0<C> * 1.0<F>

    let toRankine (x: float<C>) =
      ((x + 273.15<C>) * 1.8) / 1.0<C> * 1.0<R>

  [<RequireQualifiedAccess>]
  module Fahrenheit =

    let create (x: float) = x * 1.0<F>

    let toKelvin (x: float<F>) =
      ((x - 32.0<F>) * (5.0 / 9.0) + 273.15<F>) / 1.0<F> * 1.0<K>

    let toCelsius (x: float<F>) =
      ((x - 32.0<F>) * (5.0 / 9.0)) / 1.0<F> * 1.0<C>

    let toRankine (x: float<F>) = (x + 459.67<F>) / 1.0<F> * 1.0<R>

  [<RequireQualifiedAccess>]
  module Rankine =

    let create (x: float) = x * 1.0<R>
    let toKelvin (x: float<R>) = (x * (5.0 / 9.0)) / 1.0<R> * 1.0<K>

    let toCelsius (x: float<R>) =
      ((x - 491.67<R>) * (5.0 / 9.0)) / 1.0<R> * 1.0<C>

    let toFahrenheit (x: float<R>) = (x - 459.67<R>) / 1.0<R> * 1.0<F>
